src/ImageSharp/Formats/Jpeg/Components/Decoder/ArithmeticScanDecoder.cs

```diff
@@ -0,0 +1,1250 @@
+            // According to libjpeg the range covers 8bit only?
+            if (this.SuccessiveLow > 13)
+            {
+                invalid = true;
+            }
+
+            if (invalid)
+            {
+                JpegThrowHelper.ThrowBadProgressiveScan(this.SpectralStart, this.SpectralEnd, this.SuccessiveHigh, this.SuccessiveLow);
+            }
+        }
+
+        private void ParseBaselineDataInterleaved()
+        {
+            int mcu = 0;
+            int mcusPerColumn = this.frame.McusPerColumn;
+            int mcusPerLine = this.frame.McusPerLine;
+            ref JpegBitReader reader = ref this.scanBuffer;
+
+            for (int j = 0; j < mcusPerColumn; j++)
+            {
+                this.cancellationToken.ThrowIfCancellationRequested();
+
+                // Decode from binary to spectral.
+                for (int i = 0; i < mcusPerLine; i++)
+                {
+                    // Scan an interleaved mcu... process components in order.
+                    int mcuCol = mcu % mcusPerLine;
+                    for (int k = 0; k < this.scanComponentCount; k++)
+                    {
+                        int order = this.frame.ComponentOrder[k];
+                        var component = this.components[order] as ArithmeticDecodingComponent;
+
+                        ref ArithmeticDecodingTable dcDecodingTable = ref this.dcDecodingTables[component.DcTableId];
+                        ref ArithmeticDecodingTable acDecodingTable = ref this.acDecodingTables[component.AcTableId];
+
+                        int h = component.HorizontalSamplingFactor;
+                        int v = component.VerticalSamplingFactor;
+
+                        // Scan out an mcu's worth of this component; that's just determined
+                        // by the basic H and V specified for the component.
+                        for (int y = 0; y < v; y++)
+                        {
+                            Span<Block8x8> blockSpan = component.SpectralBlocks.DangerousGetRowSpan(y);
+                            ref Block8x8 blockRef = ref MemoryMarshal.GetReference(blockSpan);
+
+                            for (int x = 0; x < h; x++)
+                            {
+                                if (reader.NoData)
+                                {
+                                    // It is very likely that some spectral data was decoded before we've encountered 'end of scan'
+                                    // so we need to decode what's left and return (or maybe throw?)
+                                    this.spectralConverter.ConvertStrideBaseline();
+                                    return;
+                                }
+
+                                int blockCol = (mcuCol * h) + x;
+
+                                this.DecodeBlockBaseline(
+                                    component,
+                                    ref Unsafe.Add(ref blockRef, blockCol),
```

Comment:
@gfoidl I dont understand the suggestion: How can I use `uint` here, when `Unsafe.Add` expects `int`?

Reply:
Here it's for the overload `Unsafe.Add<T>(ref T source, nuint elementOffset)` and uses the implicit conversion from `uint` -> `nuint`.
So the long form of the suggestion is
 ```suggestion
                                    ref Unsafe.Add(ref blockRef, (nuint)(uint)blockCol),
```
but as the compiler does the second conversion for us, I've shortened it to
```suggestion
                                    ref Unsafe.Add(ref blockRef, (uint)blockCol),
```

---

In essence, and to avoid the `movsxd` (the sign extending move), we need to convert the `int` to any of `IntPtr` / `nint` or `UIntPtr` / `nuint`. As first step here is to prove to the JIT that there's only positive numbers by the `uint`-cast, which elides the `movsxd` emitting. 
The actual target type (`nint` or `nuint` respectively there IntPtr base variants) doesn't really matter then.

Reply:
Thank you for the explanation. Now I see why it does not work: The uint overload is only available on .net6.0, so cant use that here.

Reply:
Ah, sorry, didn't check that. The use `(nint)(uint)` for the cast.